node/flows.re

```diff
@@ -294,20 +294,72 @@ let received_signature = (state, update_state, ~hash, ~signature) => {
 let received_operation =
     (state, update_state, request: Networking.Operation_gossip.request) =>
   if (!List.mem(request.operation, state.Node.pending_side_ops)) {
-    Lwt.async(() => {
-      let _state =
-        update_state(
-          Node.{
-            ...state,
-            pending_side_ops: [
-              request.operation,
-              ...state.Node.pending_side_ops,
-            ],
-          },
+    let {
+      signature,
+      nonce,
+      block_height,
+      Operation.Side_chain.source,
+      kind,
+      hash,
+    } =
+      request.operation;
+    Operation.Side_chain.(
+      switch (request.operation.kind) {
+      | Add_validator(_)
+      | Remove_validator(_) =>
+        let.assert () = (
+          `Invalid_signature_author,
+          Address.compare(
+            state.Node.identity.t,
+            Signature.public_key(
+              request.operation.Operation.Side_chain.signature,
+            ),
+          )
+          == 0,
         );
-      let.await () = Networking.broadcast_operation_gossip(state, request);
-      Lwt.return();
-    });
+        switch (
+          Operation.Side_chain.verify(
+            ~hash,
+            ~signature,
+            ~nonce,
+            ~block_height,
+            ~source,
+            ~kind,
+          )
+        ) {
+        | Ok(_) =>
+          let _: State.t =
+            update_state(
+              Node.{
+                ...state,
+                pending_side_ops: [
+                  request.operation,
+                  ...state.Node.pending_side_ops,
+                ],
+              },
+            );
+          Ok();
+        | Error(_) => Error(`Failed_to_verify_payload)
+        };
+      | _ =>
+        Lwt.async(() => {
+          let _state =
+            update_state(
+              Node.{
+                ...state,
+                pending_side_ops: [
+                  request.operation,
+                  ...state.Node.pending_side_ops,
+                ],
+              },
+            );
+          Networking.broadcast_operation_gossip(state, request);
+        });
+        Ok();
+      }
+    );
+  } else {
+    Ok();
   };
```

Comment:
Please change this using `let.ok` for early return, there is duplicated logic and the code is really nested.